config/config_file.go

```diff
@@ -1,109 +1,84 @@
 package config
 
 import (
+	"context"
 	"fmt"
+	"io/ioutil"
 	"os"
-	"os/user"
 	"path/filepath"
 
+	"go4.org/xdgdir"
 	"gopkg.in/ini.v1"
+	"zombiezen.com/go/log"
 )
 
-func configFilePath() (string, error) {
-	u, err := user.Current()
-
-	if err != nil {
-		return "", err
-	}
-
-	configDir := filepath.Join(u.HomeDir, ".config", "yb")
-	mkdirAsNeeded(configDir)
-	iniPath := filepath.Join(configDir, "settings.ini")
+const (
+	dirName      = "yb"
+	settingsName = "settings.ini"
+)
 
-	if !pathExists(iniPath) {
-		emptyFile, _ := os.Create(iniPath)
-		emptyFile.Close()
+func loadConfigFiles() (*ini.File, error) {
+	// Gather INI files in reverse order because earlier entries should override
+	// later entries.
+	searchPaths := xdgdir.Config.SearchPaths()
+	var iniFiles []interface{}
+	for i := len(searchPaths) - 1; i >= 0; i-- {
+		data, err := ioutil.ReadFile(filepath.Join(searchPaths[i], dirName, settingsName))
+		if err != nil && !os.IsNotExist(err) {
+			log.Warnf(context.TODO(), "Load configuration: %v", err)
+			continue
+		}
+		iniFiles = append(iniFiles, data)
 	}
-
-	return iniPath, nil
-}
-
-func loadConfigFile() (*ini.File, error) {
-	iniPath, err := configFilePath()
-	if err != nil {
-		return nil, fmt.Errorf("Couldn't determine config file path: %v", err)
+	if len(iniFiles) == 0 {
+		iniFiles = append(iniFiles, []byte(nil))
 	}
 
-	cfg, err := ini.Load(iniPath)
-
+	cfg, err := ini.Load(iniFiles[0], iniFiles[1:]...)
 	if err != nil {
-		fmt.Printf("Fail to read file: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("load configuration: %w", err)
 	}
-
 	return cfg, nil
-
 }
 
-// To break an import cycle
-func pathExists(path string) bool {
-	if _, err := os.Lstat(path); os.IsNotExist(err) {
-		return false
+func resolveSectionName(cfg *ini.File, name string) string {
+	if name == "defaults" {
+		return name
 	}
-
-	return true
-}
-
-// To break an import cycle
-func mkdirAsNeeded(dir string) error {
-	if _, err := os.Stat(dir); os.IsNotExist(err) {
-		if err := os.MkdirAll(dir, 0700); err != nil {
-			return fmt.Errorf("Unable to create dir: %v", err)
-		}
+	if profile := getProfile(cfg); profile != "" {
+		return profile + ":" + name
 	}
-
-	return nil
+	return name
 }
 
-func SectionPrefix() string {
-	profile := YourBaseProfile()
-	if profile != "" {
-		return fmt.Sprintf("%s:", profile)
+func Get(section string, key string) (string, error) {
+	cfg, err := loadConfigFiles()
```

Comment:
Are we reading from disk every time `Get` is called? If so, can we cache the read in memory?

This seems unchanged from current behavior so I'd be fine if it's in a separate PR.

Reply:
Sadly, yes, it is making syscalls every time. I would like to cache the read in-memory, but I was worried in general about the interaction with `Set()`, even though the calls I've seen so far seem okay. I'm relying on the fact that the files are small enough that the OS will likely keep them in cache rather than actually hit disk to handle the overall synchronization.

Opened ch-2693 to track cleanup.